File: Data_Structure/DS_Using_Python/Full_Binary_Tree.py
""" Construct below tree
             1
           /   \
          /     \
         2       3
        / \     
       4   5  
    """

class FullBinaryTree:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None
	
	#Check if the tree is full
	def isFullBinaryTree(self, tree):
		if tree is None:
			return True
		if tree.left is None and tree.right is None:
			return True
		if tree.left and tree.right:
			return (self.isFullBinaryTree(tree.left) and self.isFullBinaryTree(tree.right))
		return False
		
	#insert data in left child node
	def insertLeft(self, newNode):
		if self.left is None:
			self.left = FullBinaryTree(newNode)
		else:
			tree = FullBinaryTree(newNode)
			tree.left = self.left
			self.left = tree
		
	#insert data in right child node
	def insertRight(self, newNode):
		if self.right is None:
			self.right = FullBinaryTree(newNode)
		else:
			tree = FullBinaryTree(newNode)
			tree.right = self.right
			self.right = tree
		
root = FullBinaryTree(1)
root.insertLeft(2)
root.insertRight(3)
root.insertLeft(4)
root.insertRight(5)

if(root.isFullBinaryTree(root)):
	print("It is a Full Binary Tree")
else:
	print("It is not a Full Binary Tree")

File: Data_Structure/Trees/Full_Binary_Tree.c
#include<stdio.h>
#include<stdbool.h>

struct node
{
    int data;
    struct node* left;
    struct node* right;
};

//Creating a new node
struct node* createNode(int item)
{
    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    newNode -> data = item;
    newNode -> left = NULL;
    newNode -> right = NULL;
    return newNode;
}

//Check if the tree is full
bool isFullBinaryTree(struct node* root)
{
    if(root == NULL)
    {
        return true;
    }
    if(root -> left == NULL && root -> right == NULL)
    {
        return true;
    }
    if((root -> left) && (root -> right))
    {
        return(isFullBinaryTree(root -> left) && isFullBinaryTree(root -> right));
    }
    return false;
}

//Insert data in left child node
struct node* insertLeft(struct node* root, int item)
{
    root -> left = createNode(item);
    return root -> left;
}

//Insert data in right child node
struct node* insertRight(struct node* root, int item)
{
    root -> right = createNode(item);
    return root -> right;
}

int main()
{
    struct node* root = createNode(2);
    insertLeft(root, 5);
    insertRight(root, 7);
    insertLeft(root -> left, 8);
    insertRight(root -> left, 82);
    
    if(isFullBinaryTree(root))
    {
        printf("It is a Full Binary Tree");
    }
    else
    {
        printf("It is not a Full Binary Tree");
    }
    
    return 0;
}